homeworks/08-functional-reactive-programming/src/index.ts

```diff
@@ -0,0 +1,58 @@
+import {animationFrameScheduler, combineLatest, fromEvent, interval, Observable, timer} from "rxjs"
+import {CellSize, draw} from "./drawing"
+import {Board, ColRow, Coords, Score} from "./domain"
+import {hasAnimal, InitialBoard, moveAnimal} from "./game-logic"
+import {startWith, map, scan, share} from "rxjs/operators"
+
+function runGame(canvas: HTMLCanvasElement, context: CanvasRenderingContext2D): void {
+    const mouseClicks$: Observable<ColRow | null> = fromEvent<MouseEvent>(canvas, "click")
+        .pipe(
+            map((e) => {
+                const col = Math.floor(e.x / CellSize)
+                const row = Math.floor(e.y / CellSize)
+                return { column: col, row: row } as ColRow
+            }),
+            startWith(null),
+        )
+
+    const mousePositions$: Observable<Coords | null> = fromEvent<MouseEvent>(canvas, "mousemove")
+        .pipe(
+            map((e) => ({ x: e.x, y: e.y } as Coords)),
+            startWith(null),
+        )
+
+    // regular moving of animals from one window to the other
+    const board$: Observable<Board> = timer(0, 1000)
+        .pipe(
+            scan((board) => moveAnimal(board), InitialBoard),
+            share(),
+        )
+
+    const frames$: Observable<number> = interval(0, animationFrameScheduler)
+
+    const score$: Observable<Score> = combineLatest([board$, mouseClicks$])
+        .pipe(
+            map(([board, click]) => click ? (hasAnimal(board, click) ? 1 : 0) : 0),
+            scan((acc, v) => acc + v as Score, 0 as Score),
+        )
+
+    combineLatest([score$, mousePositions$, board$, frames$])
```

Comment:
I think `frames$` stream is unneeded here. You don't have any things you need to change over time (except cats position every second) so this stream will force a browser to call draw function each 16.67ms (ideally) even when a user doesn't move a mouse.
Re-drawing could be triggered when: 
- mouse position is changed
- each second (cats changes its position)
- a user clicks on a board.

Reply:
I agree. It's more "future proofing" in case we are adding any time-dependent animations or FPS counter.